[lit] Echo full RUN lines in case of external shells

jdenny-ornl opened the pull request:
Before <https://reviews.llvm.org/D154984> and <https://reviews.llvm.org/D156954>, lit reported full RUN lines in a `Script:` section.  Now, in the case of lit's internal shell, it's the execution trace that includes them.  However, if lit is configured to use an external shell (e.g., bash, windows `cmd`), they aren't reported at all.

A fix was requested at the following:

* <https://reviews.llvm.org/D154984#4627605>
* <https://discourse.llvm.org/t/rfc-improving-lits-debug-output/72839/35?u=jdenny-ornl>

This patch does not correctly address the case when the external shell is windows `cmd`.  As discussed at <https://github.com/llvm/llvm-project/pull/65242>, it's not clear whether that's a use case that people still care about, and it seems to be generally broken anyway.

RoboTux commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -1153,8 +1153,14 @@ def executeScript(test, litConfig, tmpBase, commands, cwd):
+                commands[j] += f": {shlex.quote(command.lstrip())} >&2 " \
+                               f"&& {command}"
+            else:
+                commands[j] += " has no command after substitutions"
```
Could you add a testcase for this codepath? Thanks!

jdenny-ornl replied:
Good idea, especially given that I forgot the stderr redirection in that case.  I pushed a commit.

banach-space reviewed (approved):
LGTM, thank you!

mikaelholmen commented:
I like this. Thank you!

RoboTux reviewed (approved):
LGTM too.

jdenny-ornl committed:
[lit] Echo full RUN lines in case of external shells

Before <https://reviews.llvm.org/D154984> and
<https://reviews.llvm.org/D156954>, lit reported full RUN lines in a
`Script:` section.  Now, in the case of lit's internal shell, it's the
execution trace that includes them.  However, if lit is configured to
use an external shell (e.g., bash, windows `cmd`), they aren't
reported at all.

A fix was requested at the following:

* <https://reviews.llvm.org/D154984#4627605>
* <https://discourse.llvm.org/t/rfc-improving-lits-debug-output/72839/35?u=jdenny-ornl>

This patch does not correctly address the case when the external shell
is windows `cmd`.  As discussed at
<https://github.com/llvm/llvm-project/pull/65242>, it's not clear
whether that's a use case that people still care about, and it seems
to be generally broken anyway.

jsji commented:
Looks like this PR might cause unexpected failure in testing, and kill buildbot worker (also python code). Examples can be found: https://lab.llvm.org/buildbot/#/builders/184/builds/6236.
https://lab.llvm.org/buildbot/#/builders/197/builds/9358

jdenny-ornl commented:
I pushed a revert (efec733bf5bb97b34361c4ce49346edc6afa3866).  I cannot tell from the logs what happened.

jdenny-ornl commented:
@jsji Thanks for alerting me to the bot fails.  The revert seems to have fixed at list one bot.  Did you see any additional clue in the logs about what happened?  All I see is:

```
remoteFailed: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionLost'>: Connection to the other side was lost in a non-clean fashion.
]
```

 I don't know how to debug that.

jsji commented:
Sorry, no idea about why either.

dyung commented:
It's probably not much help, but this affected one of my build bots, and when I ran check-all manually on the machine, it died with an error "User defined signal 2":
```
PASS: libFuzzer-x86_64-default-Linux :: recommended-dictionary.test (72990 of 75935)
PASS: libFuzzer-x86_64-default-Linux :: cleanse.test (72991 of 75935)                                                                                                                                                                             
User defined signal 2
```

petrhosek commented:
We saw the same issue on our bots as well, see for example:

https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8770760294040862593/overview

We did some investigation but haven't found the culprit either, but we saw trap faults in the kernel log.

jdenny-ornl commented:
Thanks for the info, and sorry for all the trouble.  I am now able to reproduce on a local system: `ninja check-fuzzer` kills my terminal window with signal 12.  I'll keep poking at it as I have time.  If you see any additional clues, please post them here.

uweigand commented:
I'm running into this issue on the s390x builder as well.  It seems the problem happens in the ./compiler-rt/test/fuzzer/fork-sigusr.test case here:
```
RUN: %run %t/ForkSIGUSR -fork=3 -ignore_crashes=1 2>%t/log & export PID=$!
RUN: sleep 3
RUN: kill -SIGUSR2 $PID
RUN: sleep 6
RUN: cat %t/log | FileCheck %s --dump-input=fail
```

Normally, the PID variable is set to the pid of the ForkSIGUSR process, which then gets killed with SIGUSR2 and terminates cleanly.

However, with this patch applied, the PID variable get set to the pid of a *shell* process executing
```
  echo "..." && ../ForkSIGUSR ...
```
The shell eats the SIGUSR2 and prints a message, but the ForkSIGUSR process never gets the signal and keeps running indefinitely.

This somehow causes the buildbot connection to terminate - not sure exactly why but possibly some bad reaction to that shell getting the signal?

Unfortunately whenever I re-start the build bot, it thinks it need to re-try the previous build since that had the failed connection - but that previous build still has this patch applied and therefore terminates again :-(   I'm trying to work around this now.

jdenny-ornl commented:
@uweigand Thanks!  That was the clue I needed.

I found there are four tests that have a similar pattern and cause trouble on my local system:

- fork-sigusr.test: Seemed to hang, but I never saw it kill the shell that was running check-fuzzer.  That appears to be a different result than @uweigand saw, so maybe there's something non-deterministic, or maybe our shells are different.
- merge-sigusr.test: Killed the shell.
- sigint.test: Seemed to hang.
- sigusr.test: Killed the shell.

If I disable those tests, check-fuzzer behaves as at the parent commit.

I have a simple lit fix I've pushed to this branch.  It also makes check-fuzzer behave as at the parent commit.

I'm hesitant to land the new version as last time was a bit of a disaster.  Based on @uweigand's comments,  some bots are still recovering.  Sorry!  Who knows what other trouble might be lurking on the next attempt.

Any advice on how to proceed?

jdenny-ornl commented:
Here's the fix: <https://github.com/jdenny-ornl/llvm-project/commit/61e272e9e20c5eda98f9132c3f24388abf443a83>

Unfortunately, I made the mistake of pushing the revert directly to git (old workflow) instead of clicking the revert button here in the PR.  I'm not sure how to reopen the PR now.  I can start a new PR if that's desirable.

dyung commented:
> Unfortunately whenever I re-start the build bot, it thinks it need to re-try the previous build since that had the failed connection - but that previous build still has this patch applied and therefore terminates again :-( I'm trying to work around this now.

@uweigand if you are still trying to recover your buildbot, this is the process I used which I was able to recover mine. Thanks to the info everyone found here, it seems there are 4 affected tests fork-sigusr.test, merge-sigusr.test, sigint.test and sigusr.test. With the worker offline, I manually edited the files to add a `, *` to the `UNSUPPORTED` line to get the run to skip the tests so it wouldn't fail. I then stashed the change and brought the worker back online. When the job rerun had finished the sync and was doing the build, I went to the local git repo and unstashed the test changes and then let the build/test complete. Then before the next run is executed, just revert the local test changes. The tests were not exactly skipped (they ended up being "UNRESOLVED" because of the new syntax required for the UNSUPPORTED lines), but the net result was the run completed and moved on to the next job which then passed. Hope this helps you to recover your bot.

"Completed" job: https://lab.llvm.org/buildbot/#/builders/247/builds/8680

uweigand commented:
Thanks for the tips.  I've managed to get the bot going again by clearing the build queues in the web interface, so it would reset to current mainline.   All s390x bots are now green again.

jdenny-ornl commented:
I'm trying to figure out how to resurrect this PR after rebasing onto main and applying fixes.  It will need additional review.  Is it best to just to create a new PR, or is there some way to reopen this one?

jdenny-ornl commented:
Based on the information I found, it's not possible to reopen this PR, so I created a new one: #66408
